File: nodejs/day16/index2.js
const { readFile, parseInt } = require("../lib");

const parse = s => {
  const command = s[0];
  let payload = s.slice(1);
  switch (command) {
    case "s":
      payload = parseInt(payload);
      break;
    case "x":
      payload = payload.split("/").map(parseInt);
      break;
    case "p":
      payload = payload.split("/");
      break;
    default:
      throw new Error("WTF");
  }
  return { command, payload };
};

const a = "a".charCodeAt(0);
let arr = Array.from({ length: 16 }, (_, i) => String.fromCharCode(i + a));

const input = readFile("./input")
  .split(",")
  .map(parse);

const spin = x => {
  arr = arr.slice(arr.length - x).concat(arr.slice(0, arr.length - x));
};

const swap = (x, y) => {
  const t = arr[x];
  arr[x] = arr[y];
  arr[y] = t;
};

const run = () => {
  for (const { command, payload } of input) {
    switch (command) {
      case "s":
        spin(payload);
        break;
      case "x":
        swap(...payload);
        break;
      case "p":
        swap(...payload.map(p => arr.indexOf(p)));
        break;

      default:
        throw new Error("WTF");
    }
  }
};

let setup,
  period,
  i = 0;
let seen = {
  [arr.join("")]: 0
};

while (true) {
  run();
  i++;
  const state = arr.join("");
  if (seen[state]) {
    setup = seen[state];
    period = i - setup;
    break;
  } else {
    seen[state] = i;
  }
}

const ITER = 1e9;
const iter = (ITER - setup) % period;
for (let i = 0; i < iter; i++) {
  run();
}

console.log(arr.join(""));

File: clojure/day16/index2.clj
(require '[clojure.string :as s])

(def parse-int #(Integer/parseInt %))

(defn split [p s] (s/split s p))

(defn cnct [& as] (into [] (apply concat as)))

(defn parse [s]
  (let [cmd (first s) payload (subs s 1)]
    (case cmd
      \s [cmd (parse-int payload)]
      \x [cmd (->> payload (split #"/") (map parse-int) (into []))]
      \p [cmd (->> payload (split #"/") (map first) (into []))]
      (throw (ex-info "WTF" {:cmd s})))))

(def input (->>
  "./input"
  slurp
  (split #",")
  (map parse)))

(defn spin [arr x]
  (let [
    L (count arr)
    x' (- L x)
    a1 (subvec arr x')
    a2 (subvec arr 0 x')
  ] (cnct a1 a2)))

(defn swap [arr [x y]]
  (let [
    t (get arr x)
    arr' (assoc arr x (get arr y))
  ] (assoc arr' y t)))

(defn swap-vals [arr [a b]] 
  (let [
    x (.indexOf arr a)
    y (.indexOf arr b)
  ] (swap arr [x y])))

(def arr
  (let [
    L 16
    a (int \a)
    r (range L)
    r' (map #(char (+ a %)) r)
  ] (into [] r')))

(defn run [arr]
  (loop [
    arr arr
    rem-cmd input
  ] (if (== 0 (count rem-cmd)) arr
      (let [
        [cmd payload] (first rem-cmd)
        rem-cmd' (rest rem-cmd)
        op (case cmd
          \s spin
          \x swap
          \p swap-vals)
        arr' (op arr payload)
      ] (recur arr' rem-cmd')))))

(def period-info
  (loop [
    arr arr
    seen {arr 0}
    i 0
  ] (let [
      i' (inc i)
      arr' (run arr)
    ] (if (contains? seen arr')
        [arr' (get seen arr') (- i' (get seen arr'))]
        (recur arr' (assoc seen arr' i') i')))))

(def ITER 1e9)

(def res (let [
    [arr' setup period] period-info
    iter (mod (- ITER setup) period)
  ]
    (loop [
      arr arr'
      i 0
    ] (if (== iter i) arr
        (recur (run arr) (inc i))))))

(println (s/join "" res))